internal/characters/xianyun/asc.go

```diff
@@ -0,0 +1,126 @@
+package xianyun
+
+import (
+	"github.com/genshinsim/gcsim/pkg/core/attacks"
+	"github.com/genshinsim/gcsim/pkg/core/attributes"
+	"github.com/genshinsim/gcsim/pkg/core/combat"
+	"github.com/genshinsim/gcsim/pkg/core/event"
+	"github.com/genshinsim/gcsim/pkg/core/glog"
+	"github.com/genshinsim/gcsim/pkg/core/player/character"
+	"github.com/genshinsim/gcsim/pkg/core/targets"
+	"github.com/genshinsim/gcsim/pkg/modifier"
+)
+
+const a4ICDKey = "xianyun-a4-icd"
+const a1Key = "xianyun-a1"
+const a1Dur = 20 * 60
+
+var a1Crit = []float64{0.0, 0.04, 0.06, 0.08, 0.10}
+
+// a1: For every enemy hit by Driftcloud Wave,
+// all party members gain 1 stack of Boost, lasting 20s, max 4 stacks.
+// Boost increases Plunge DMG's Crit Rate by 4%/6%/8%/10%,
+// each stack's duration is calculated independently.
+
+func (c *char) a1() {
+	for idx, char := range c.Core.Player.Chars() {
+		mCR := make([]float64, attributes.EndStatType)
+		char.AddAttackMod(character.AttackMod{
+			Base: modifier.NewBase("xianyun-a1-buff", -1),
+			Amount: func(atk *combat.AttackEvent, t combat.Target) ([]float64, bool) {
+				if atk.Info.AttackTag != attacks.AttackTagPlunge {
+					return nil, false
+				}
+				stackCount := min(c.a1Buffer[idx], 4)
+				if stackCount == 0 {
+					return nil, false
+				}
+				mCR[attributes.CR] = a1Crit[stackCount]
+				return mCR, true
+			},
+		})
+	}
+}
+
+func (c *char) a1cb() combat.AttackCBFunc {
+	if c.Base.Ascension < 1 {
+		return nil
+	}
+	return func(a combat.AttackCB) {
+		if a.Target.Type() != targets.TargettableEnemy {
+			return
+		}
+
+		for i, char := range c.Core.Player.Chars() {
+			idx := i
+			c.a1Buffer[idx] += 1
+			char.AddStatus(a1Key, a1Dur, true)
+			char.SetTag(a1Key, min(c.a1Buffer[idx], 4))
+			char.QueueCharTask(func() {
+				// tags currently aren't visible in the results UI
+				// the user can still access it using .char.tags.xianyun-a1
+				c.a1Buffer[idx] -= 1
+				char.SetTag(a1Key, min(c.a1Buffer[idx], 4))
+
+				// TODO: Don't need to specifically remove A1 Status because it should have expired at the same time
+				char.DeleteStatus(a1Key)
+			}, a1Dur)
+		}
+	}
+}
+
+func (c *char) a4AtkUpdate() {
+	c.a4Atk = c.Base.Atk*(1+c.Stat(attributes.ATKP)) + c.Stat(attributes.ATK)
```

Comment:
you'll want to use `.Stats` instead here because each `.Stat` call has to loop all the mods so here it's doing it twice.